README.markdown
# grunt-integration [![Build Status](https://travis-ci.org/bevacqua/grunt-integration.png?branch=master)](https://travis-ci.org/bevacqua/grunt-integration) [![Dependency Status](https://gemnasium.com/bevacqua/grunt-integration.png)](https://gemnasium.com/bevacqua/grunt-integration) [![help me on gittip](http://gbindex.ssokolow.com/img/gittip-43x20.png)](https://www.gittip.com/bevacqua/) [![flattr.png](https://api.flattr.com/button/flattr-badge-large.png)](https://flattr.com/submit/auto?user_id=nzgb&url=https%3A%2F%2Fgithub.com%2Fbevacqua%2Fgrunt-integration)

> Run Integration Tests using Selenium, Mocha, a Server, and a Browser

### Features

These are the goals of `grunt-integration`.

- Start a local selenium server instance
- Start a local program, such as `node` application
- Wait for the program to listen on a specific port
- Execute integration tests using [Mocha][1] and [WebDriver][2]
- Using real browsers, such as Chrome
- Automatically, in one command
- Less painful installation process, please!

[**Read the article on Pony Foo**][3]

### Installation

You'll need `node`, `java`. If you have `apt-get`, you can get `java` using the command below. If you like clicking on things, just google for a jdk distribution, and install that.

```shell
sudo apt-get install openjdk-7-jdk -y
```

Then install `grunt-integration` from `npm`.

```shell
npm install --save-dev grunt-integration
```

### Configuration

Here are the default `options`.

Option              | Default                           | Details
--------------------|-----------------------------------|------------------------
`program`           | `'node app'`                      | This will be spawned and expected to listen
`program_port`      | `process.env.TEST_PORT` or `3333` | Which port will you use?
`program_inherit`   | `true`                            | Will program output be printed to the terminal?
`selenium_inherit`  | `false`                           | Will selenium server output be printed to the terminal?
`tests`             | See below                         | These are the options passed to `mocha`

Here are the default `tests` options. Note that if you want to change something in `options.tests`, you'll need to replace the whole object, so just copy and paste the defaults, and work from there. That's just how `this.options()` works in Grunt at the moment.

```js
tests: {
    hostname: '127.0.0.1',
    port: 4444,
    usePromises: true,
    concurrency: 1,
    browsers: [{ browserName: 'chrome' }],
    timeout: 40000,
    reporter: 'spec'
}
```

#### These defaults are good enough that you don't need to touch any of it to get started.

You can just configure it like this to get started.

```js
grunt.initConfig({
    integration: {
        src: ['test/integration/**/*.js']
    }
});
```

### Usage

For usage examples, you're going to have to sink your teeth in [wd][2].

### License

MIT

  [1]: https://github.com/visionmedia/mocha
  [2]: https://github.com/admc/wd
  [3]: http://blog.ponyfoo.com/2013/12/20/is-webdriver-as-good-as-it-gets "Is WebDriver as good as it gets?"


CHANGELOG.md
# 0.1.0 Sauce

- Initial Release


lib/mocha.js
var Mocha = require('mocha');
var path = require('path');
var Module = require('module');
var fs = require('fs');
var path = require('path');
var domain = require('domain');
var grunt = require('grunt');

module.exports = function (opts, fileGroup, browser, done) {
    var mocha = new Mocha(opts);

    mocha.suite.on('pre-require', function () {
        this.ctx.browser = browser;
        this.ctx.wd = opts.wd;
    });

    grunt.file.expand({ filter: 'isFile' }, fileGroup.src).forEach(function (file) {
        var mod = path.resolve(file);
        if (Module._cache[mod]) {
            delete Module._cache[mod];
        }
        mocha.addFile(mod);
    });

    if (mocha.files.length) {
        mocha.loadFiles();
    }

    var context = domain.create();
    var mochaOptions = mocha.options;
    var mochaRunner = new Mocha.Runner(mocha.suite);
    var mochaReporter = new mocha._reporter(mochaRunner);

    mochaRunner.ignoreLeaks = mochaOptions.ignoreLeaks !== false;
    mochaRunner.asyncOnly = mochaOptions.asyncOnly;

    if (mochaOptions.grep) {
        mochaRunner.grep(mochaOptions.grep, mochaOptions.invert);
    }

    if (mochaOptions.globals) {
        mochaRunner.globals(mochaOptions.globals);
    }

    context.on('error', mochaRunner.uncaught.bind(mochaRunner));
    context.run(function () {
        mochaRunner.run(function (errCount) {
            var err;
            if (errCount !== 0) {
                err = new Error('Tests encountered ' + errCount + ' errors.');
            }
            done(err);
        });
    });
};


lib/testdrive.js
'use strict';

var wd = require('wd');
var chalk = require('chalk');
var async = require('async');
var mocha = require('./mocha.js');

module.exports = function (files, opts, done) {

    var testResult;

    async.forEachSeries(files, function (fileGroup, next) {
        var queue = async.queue(function (browserOpts, dequeue) {
            prepare(browserOpts, fileGroup, dequeue);
        }, opts.concurrency);

        opts.wd = wd;
        opts.browsers.forEach(function (browserOpts) {
            queue.push(browserOpts, function (err) {
                if (err) {
                    testResult = new Error('One or more tests on Selenium failed.');
                }
            });
        });

        queue.drain = function () {
            next(testResult);
        };

    }, done);

    function prepare(browserOpts, fileGroup, next) {
        var style = opts.usePromises ? 'promiseChainRemote': 'remote';
        var browser = wd[style]({
            hostname: opts.hostname,
            port: opts.port
        });
        browserOpts.browserTitle = browserOpts.browserName;
        browser.browserTitle = browserOpts.browserTitle;
        browser.init(browserOpts, function (err) {
            if (err) {
                return next(err);
            }
            console.log(chalk.magenta('Running tests on %s'), browser.browserTitle);

            mocha(opts, fileGroup, browser, function (err) {
                browser.quit();
                next(err);
            });
        });
    }
};


tasks/integration.js
'use strict';

module.exports = function (grunt) {
    var _ = require('lodash');
    var async = require('async');
    var chalk = require('chalk');
    var spawn = require('spawn-machine');
    var finder = require('process-finder');
    var pkg = require('../package.json');
    var context = {};

    grunt.registerMultiTask('integration', pkg.description, function () {
        var done = this.async();
        var name = this.target;

        var ctx = context[name] = {
            name: name,
            options: this.options({
                program: 'node app',
                program_port: process.env.TEST_PORT || 3333,
                program_inherit: true,
                selenium_inherit: false,
                tests: {
                    hostname: '127.0.0.1',
                    port: 4444,
                    usePromises: true,
                    concurrency: 1,
                    browsers: [{ browserName: 'chrome' }],
                    timeout: 40000,
                    reporter: 'spec'
                }
            }),
            files: this.files
        };

        async.series([
            async.apply(selenium, ctx),
            async.apply(program, ctx),
            async.apply(tests, ctx)
        ], done);
    });

    function selenium (ctx, next) {
        var standalone = require('../lib/standalone.js');
        var rready = /Started org\.openqa\.jetty\.jetty\.Server/i;
        var rwarning = /\b(warn(ing)?|error|exception)\b/i;

        console.log(chalk.magenta('Spawning selenium server...'));

        var child = standalone.start({ stdio: 'pipe' });

        ctx.selenium = {
            name: 'Selenium',
            child: child
        };

        var teardown = cleanup.bind(null, ctx.selenium);

        child.stdout.on('data', function (data) {
            var stdout = data.toString();
            if (ctx.options.selenium_inherit) {
                process.stdout.write(stdout);
            }

            if (rready.test(stdout)) {
                console.log(chalk.magenta('Selenium server operational'));
                next();
            }
        });

        child.stderr.on('data', function (data) {
            if (ctx.options.selenium_inherit) {
                process.stderr.write(data);
            } else {
                var print = rwarning.test(data);
                if (print) {
                    process.stderr.write(chalk.red('[SELENIUM] ' + data));
                }
            }
        });

        process.on('exit', teardown);
    }


    function program (ctx, next) {
        var port = ctx.options.program_port;
        var watcher = finder.watch({ port: port, frequency: 400 });
        var env = _.clone(process.env);
        env.PORT = port;

        console.log(chalk.magenta('Spawning application...'));

        var child = spawn(ctx.options.program, {
            stdio: ctx.options.program_inherit ? 'inherit' : 'pipe',
            env: env
        });

        ctx.app = {
            name: 'Application',
            child: child
        };

        var teardown = cleanup.bind(null, ctx.app);

        watcher.on('listen', function () {
            console.log(chalk.magenta('Application online'));
            next();
        });

        process.on('exit', teardown);
    }

    function tests (ctx, next) {
        var testdrive = require('../lib/testdrive.js');

        console.log(chalk.magenta('Preparing to execute tests...'));
        testdrive(ctx.files, ctx.options.tests, next);
    }

    function cleanup (target) {
        if (target.terminated) {
            return;
        }

        target.terminated = true;

        if (target.child) {
            target.child.kill('SIGHUP');
            console.log('%s process %s shutting down', chalk.magenta(target.name), chalk.yellow(target.child.pid));
        } else {
            console.log(chalk.magenta(target.name), chalk.red('process not found'));
        }
    }
};
